### PR TITLE
make it possible to serialize SnoopCompile results

### PR DESCRIPTION
**Pull request by JeffBezanson**
Addresses #39655. I don't think it will realistically be possible for JLD to support these data structures, but `Serialization` might as well support it at least.

- use `nothing` instead of NULL as a sentinel in specializations. This field is technically visible and we don't support user-visible SimpleVectors with NULLs
- remove some Serializer restrictions around saving Methods. The code path for `method.generator` was broken. We disallowed serializing non-toplevel MethodInstance objects, but I don't see a good reason for that --- we might as well just save the data even if you can't do much with it.

### CODE REVIEW

**Review by Sacha0 — APPROVED** [review]
What I follow looks good. Thanks Jeff! :)

**Comment by Sacha0** [comment]
Backport candidate, or bridge too far? :)

**Comment by JeffBezanson** [comment]
The serializer part I think is just a bugfix, since the `:generator` code path only errored before (`generator.inferred` doesn't exist anymore). The other change is internal, but could affect packages that look at the `specializations` array. That may be ok though.

File: src/gf.c [hunk 1]
```diff
@@ -146,19 +146,20 @@ JL_DLLEXPORT jl_method_instance_t *jl_specializations_get_linfo(jl_method_t *m J
...
             jl_method_instance_t *mi = jl_get_specialized(m, type, sparams);
             JL_GC_PUSH1(&mi);
-            if (hv ? (i + 1 >= cl || jl_svecref(specializations, i + 1) != NULL) : (i <= 1 || jl_svecref(specializations, i - 2) != NULL)) {
+            if (hv ? (i + 1 >= cl || jl_svecref(specializations, i + 1) != jl_nothing) : (i <= 1 || jl_svecref(specializations, i - 2) != jl_nothing)) {
                 size_t ncl = cl < 8 ? 8 : (cl*3)>>1;
                 jl_svec_t *nc = jl_alloc_svec_uninit(ncl);
                 if (i > 0)
                     memcpy((char*)jl_svec_data(nc), jl_svec_data(specializations), sizeof(void*) * i);
-                memset((char*)jl_svec_data(nc) + sizeof(void*) * i, 0, sizeof(void*) * (ncl - cl));
+                for(int j = 0; j < ncl - cl; j++)
```
Comment (vtjnash) [inline src/gf.c:161]:
```suggestion
                for (int j = 0; j < ncl - cl; j++)
```

**Comment by vtjnash** [comment]
This seems generally somewhat risky to copy in this way, but generally no worse than deepcopy of other things

**Pull request merged**